---
title: "pergene_sigfit_1Mb_min10percent_IgDenovo"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache=TRUE, fig.path="./graphics/plot",autodep=TRUE,fig.width=5, fig.height=4)
knitr::opts_knit$set(root.dir = '/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/genomic_distribution_vs_pcawg/pergene_sigfit_1Mb_min10percent_IgDenovo')
library(ggplot2)
library(cowplot)
library(caTools)
library(reshape2)
library(dplyr)
library(hdp)
library(RColorBrewer)
library(grid)
library(gridExtra)
library(ggpubr)
library(plyr)
#location="farm"
location="local"

if (location=="farm"){
  root="/lustre/scratch116/casm/cgp/users/hm8"
} else if (location=="local"){
  root="/Users/hm8/volumes/hm8_network"
} else {
  warning("location not set- using current directory")
  root=getwd()
}
```


###### label Ig regions
```{r, message=FALSE, warning=FALSE}
  # identify Ig regions (+/1 100KB)
# if either of the ends lies inside the following Ig OR
# if either of the Ig coordinates lies inside the bins
ighst = 106304735-100000 # 14
ighend = 107283226+100000 # 14
iglst = 22385390-100000 # 22
iglend = 23263607+100000 # 22
igkst = 89160078-100000 # 2
igkend = 90274237+100000 # 2
cs_start = 106053274-100000 #14
cs_end = 106322322+100000 #14
annotate_vdj_vector = function(X2){
  X = X2
  X[1] = as.character(X2[1])
  X[2] = as.numeric(X2[2])
  X[3] = as.numeric(X2[3])
  if ( X[1] == "14" & ( (X[2]>=ighst & X[2]<=ighend) | (X[3]>=ighst & X[3]<=ighend) ) ) {return("igh")} else
    if ( X[1] == "22" & ( (X[2]>=iglst & X[2]<=iglend) | (X[3]>=iglst & X[3]<=iglend) ) ) {return("igl")} else
      if ( X[1] == "2" & ( (X[2]>=igkst & X[2]<=igkend) | (X[3]>=igkst & X[3]<=igkend) ) ) {return("igk")} else
        if ( X[1] == "14" & ( (X[2]>=cs_start & X[2]<=cs_end) | (X[3]>=cs_start & X[3]<=cs_end) ) ) {return("classS")} else
   if ( X[1] == "14" & ( (X[2]>=ighst & X[3]<=ighst) | (X[2]>=ighend & X[3]<=ighend) ) ) {return("igh")} else
    if ( X[1] == "22" & ( (X[2]>=iglst & X[3]<=iglst) | (X[2]>=iglend & X[3]<=iglend) ) ) {return("igl")} else
      if ( X[1] == "2" & ( (X[2]>=igkst & X[3]<=igkst) | (X[2]>=igkend & X[3]<=igkend) ) ) {return("igk")} else
        if ( X[1] == "14" & ( (X[2]>=cs_start & X[3]<=cs_start) | (X[2]>=cs_end & X[3]<=cs_end) ) ) {return("classS")} else {return(FALSE)}
}
#myVDJ = apply(data.frame(sigprop$chr, sigprop$start, sigprop$end), MARGIN=1, annotate_vdj_vector)
```

## Load in gene info
```{r, message=FALSE, warning=FALSE}
genann = read.table("/Users/hm8/sanger/genomes/human/GRCh37d5/gencode/gencode.v30lift37.basic.annotation.geneonly.genename.bed", stringsAsFactors = F)
## Also has the following extra chromosomes not in our analysis
# [25] "M"          "GL000193.1" "GL000192.1" "GL000237.1" "GL000220.1" "GL000241.1"
# [31] "GL000199.1" "GL000212.1" "GL000202.1" "GL000228.1" "GL000195.1" "GL000205.1"
# [37] "GL000204.1"
### Hugo Looked: not used
# source("/Users/hm8/sanger/genes/hugo_genes/hugoLookup3_DanL.R")
# genann_hugo = hugoLookup(genann[,4]) 
# genann[,5] = genann_hugo
# colnames(genann) = c("chr","start","end","gencode","hugo")
# save(genann, file="gencode.v30lift37.basic.annotation.geneonly.genename.hugo.bed")
```


## Per-gene mutations
```{r, message=FALSE, warning=FALSE}
mycancertypes = c("Memory_B", "CLL_mutated", "CLL_unmutated","Burkitt_lymphoma","Follicular_lymphoma","DLBC_lymphoma","Multiple_myeloma")

myatt_list = list()
for (cancertype in c(mycancertypes)){
  myatt = read.table(paste(root, "/lymphocyteWGS/mutsig_byregion/pcawg_IgDenovo/attribution_perbp_sigfit_", cancertype, "_1Mb_min10percent_IgDenovo.txt", sep=""), header=T, stringsAsFactors = F, sep="\t")
  myatt$dataset = cancertype
  myatt_samp = myatt[sample(1:nrow(myatt), size=45711),]
  myatt_list[[cancertype]] = myatt_samp[,colnames(myatt_samp) != "NA."]
}
myattDF = do.call(rbind, myatt_list)
meanS9 = mean(myattDF$SBS9, na.rm=TRUE)
meanSIg = mean(myattDF$Signature.Ig, na.rm=TRUE)
meanSblood = mean(myattDF$Signature.blood, na.rm=TRUE)
  
for (cancertype in c(mycancertypes)){
  myatt = myatt_list[[cancertype]]
  # have to add some wobble as some nucleotides have the same value
  mywobble = sample(seq(from=0.0000001, to=0.001, by=0.00000001), size=nrow(myatt), replace=FALSE)
  myatt$SBS9 = myatt$SBS9+mywobble
  myatt$Signature.Ig = myatt$Signature.Ig+mywobble
  myatt$Signature.blood = myatt$Signature.blood+mywobble

  outlist = list()
  for (i in 1:nrow(genann)){
    # add per-gene info: Nmutations, NS9, meanS9,  NSIg, meanSIg, NSblood, meanSSblood
    focal = unlist(genann[i,])
    focal_ann = subset(myatt, chr==focal[1] & pos >= as.numeric(focal[2]) & pos <= as.numeric(focal[3]) )
    
    if (nrow(focal_ann)<=1){
      outlist[[i]] = unlist(c(focal, rep(0, times=7), 0,1,0,1,0,1))
    } else {
      ## t-test- deviation from mean
      myT_S9 = t.test(x=focal_ann$SBS9,  mu=meanS9, alternative = "greater")
      t_S9 = myT_S9$statistic
      p_S9 = myT_S9$p.value
     
      myT_SIg = t.test(x=focal_ann$Signature.Ig,  mu=meanSIg, alternative = "greater")
      t_SIg = myT_SIg$statistic
      p_SIg = myT_SIg$p.value
     
      myT_Sblood = t.test(x=focal_ann$Signature.blood,  mu=meanSblood, alternative = "greater")
      t_Sblood = myT_Sblood$statistic
      p_Sblood = myT_Sblood$p.value
      outlist[[i]] = unlist(c(focal, nrow(focal_ann), sum(focal_ann$SBS9), mean(focal_ann$SBS9), sum(focal_ann$Signature.Ig), mean(focal_ann$Signature.Ig), sum(focal_ann$Signature.blood), mean(focal_ann$Signature.blood), t_S9, p_S9,t_SIg, p_SIg, t_Sblood, p_Sblood))
    } 
  }
  genann2 = data.frame(do.call(rbind, outlist))
  colnames(genann2) = c("chr","start","end","gene","N","NS9", "meanS9",  "NSIg", "meanSIg", "NSblood", "meanSblood","t_S9", "p_S9","t_SIg", "p_SIg", "t_Sblood", "p_Sblood")
  myVDJ = apply(data.frame(genann2$chr, genann2$start, genann2$end), MARGIN=1, annotate_vdj_vector)
  genann2$IG = myVDJ
  genann2$dataset = cancertype
  write.table(genann2, file=paste("S9cAIDblood_Z_per_gene_sigfit_",cancertype,"_1Mb_min10percent_IgDenovo_N45711_ttest.txt", sep=""), col.names = T, row.names = F, quote=F, sep="\t")
}
```


## Read in each dataset
```{r, message=FALSE, warning=FALSE}
mycancertypes = c("Memory_B","CLL_mutated","DLBC_lymphoma","Follicular_lymphoma","Multiple_myeloma","Burkitt_lymphoma","CLL_unmutated")
inlist = list()
for (i in 1:length(mycancertypes)){
  cancertype = mycancertypes[i]
  myinfile = read.table(file=paste("S9cAIDblood_Z_per_gene_sigfit_",cancertype,"_1Mb_min10percent_IgDenovo_N45711_ttest.txt", sep=""), header=T, stringsAsFactors = F)
  myinfile$p_S9[myinfile$p_S9==1] = NA
  myinfile$fdr_S9 = p.adjust(myinfile$p_S9, method="fdr")
  inlist[[i]] = myinfile
}
alldatasets2 = data.frame(do.call(rbind, inlist))
alldatasets2$IgTF = TRUE
alldatasets2$IgTF[alldatasets2$IG==FALSE] = FALSE
alldatasets2$dataset2 = alldatasets2$dataset
alldatasets2$dataset2 = revalue(alldatasets2$dataset2, c("Memory_B"= "Memory B", "CLL_mutated" ="CLL mutated", "DLBC_lymphoma" ="DLBC lymphoma"  ,"Follicular_lymphoma"="Follicular lymphoma", "Multiple_myeloma"="Multiple myeloma", "Burkitt_lymphoma"   ="Burkitt lymphoma","CLL_unmutated"   ="CLL unmutated" ))
mycancertypes2 = c("Memory B","CLL mutated", "DLBC lymphoma"  ,"Follicular lymphoma", "Multiple myeloma", "Burkitt lymphoma","CLL unmutated" )
```



## Create table of top genes by S9 t-test p-value
```{r, message=FALSE, warning=FALSE}
topN=5
topS9Z_focal_list = list()
topS9Z_focal_list_info = list()
# exlude CLL unmutated 
for (i in 1:6){
  focal = subset(alldatasets2, dataset==mycancertypes[i] & IgTF==FALSE)
  focal2 = focal[order(focal$p_S9, decreasing=F), ][1:topN, ]
  topS9Z_focal_list_info[[i]] = focal2
  #focal3 = focal2[focal2$ZS9>6,]
  focal3 = focal2
  topS9Z_focal_list[[i]] = focal3$gene
}
names(topS9Z_focal_list) = mycancertypes[1:6]

mysamples = unique(unlist(topS9Z_focal_list))
#"PCDH10"     "CSMD3"      "PCDH15"     "ROBO2"      "CCSER1"     "LINC02267"  "IMMP2L"    
#  [8] "AC022140.1" "AF130359.1" "CADM2"      "AC099520.1" "ROBO1"      "AC099499.1" "AC115622.1"
# [15] "AC073429.2" "ADGRL3"     "GRID2"      "AC244205.1" "FHIT"       "DGKB"       "LSAMP" 
#intersect(genes_ncAID_q01$new_gene, mysamples ) # "CADM2"      "ROBO1"      "AC244205.1" "GRID2"      "ADGRL3"

topS9Z_focalDF = subset(alldatasets2, gene %in% unique(unlist(topS9Z_focal_list)))
topS9Z_focalDF$gene = factor(topS9Z_focalDF$gene, levels=unique(unlist(topS9Z_focal_list)))
#topS9Z_focalDF$gene2 = factor(topS9Z_focalDF$gene2, levels=unique(topS9Z_focalDF$gene2))
#topS9Z_focalDF$gene2 = factor(topS9Z_focalDF$gene2, levels=unique(topS9Z_focalDF$gene2))
topS9Z_focalDF$dataset2 = factor(topS9Z_focalDF$dataset2, levels=rev(mycancertypes2))
#write.table(topS9Z_focalDF, file="dataset_genomic_clustering_bygene_ZscoreS9_top5each_N45711.txt", col.names = T, row.names = F, quote=F)
ggplot(topS9Z_focalDF,  aes(x=gene, y=dataset2, fill=-log10(p_S9) ) )+
  geom_tile()+
  scale_fill_gradientn("p (-log10)", na.value="white", guide = guide_colorbar(frame.colour = "black", ticks.colour="black"), limits=c(0,63), colours=c("white","red","darkred","black"))+  
  scale_y_discrete(expand=c(0,0)) +
  #scale_x_discrete(labels=mygenes)+
  ylab("")+
  xlab("")+
  #theme(axis.text.x = element_text(angle = 45, vjust = 1.05, hjust=1, size=8),
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1.05, size=7),
        axis.text.y = element_text(size = 8),
        legend.text=element_text(size=8),
        legend.title=element_text(size=8),
        legend.key.size = unit(0.35, 'cm'),
        panel.border = element_rect(colour = "black", fill=NA))
ggsave(file="heatmap_genomic_clustering_bygene_ZscoreS9_top5each_sigfit_1Mb_min10percent_IgDenovo_N45711_ttest.pdf", width=4.75,height=1.9)


#PCDH10, one of the non-clustered protocadherins, is identified as a tumor suppressor gene in many tumors. Recently, promoter methylation of PCDH10 was found in diffuse large B-cell lymphoma (DLBCL) but not in normal lymph nodes, suggesting that its epigenetic aberrance is essential to the lymphomagenesis. However, there are few studies on the clinicopathological relevance and prognostic significance of PCDH10 methylation status in DLBCL.
```


## Create table of top 20 genes by S9 t-test p-value
```{r, message=FALSE, warning=FALSE}
# nonIg = subset(alldatasets2, IgTF==FALSE)
# top20_S9_samples = unique(nonIg[order(nonIg$ZS9, decreasing=T),][,"gene"])[1:20]
# #top20_S9_samples = unique(alldatasets2[order(alldatasets2$ZS9, decreasing=T),][,"gene"])[1:20]
# #"PCDH10"     "CSMD3"      "IMMP2L"     "PCDH15"     "ROBO2"      "CCSER1"     "CADM2"     
# # [8] "AC115622.1" "AC073429.2" "AC099520.1" "DGKB"       "CDH12"      "ADGRL3"     "ROBO1" 
# top20_S9 = alldatasets2[alldatasets2$gene %in% top20_S9_samples, ]
# top20_S9$dataset = factor(top20_S9$dataset, levels=rev(mycancertypes))
# #intersect(genes_ncAID_q01$new_gene, top20_S9_samples) # "CADM2"  "ROBO1"  "GRID2"  "ADGRL3"
# 
# mysub = subset(top20_S9, dataset=="Memory_B")
# mygenes_top20 = mysub[order(mysub$ZS9, decreasing = T), "gene"]
# top20_S9$gene = factor(top20_S9$gene, levels=mysub[order(mysub$ZS9, decreasing = T), "gene"])
# write.table(top20_S9, file="dataset_genomic_clustering_bygene_ZscoreS9_top20_N45711.txt", col.names = T, row.names = F, quote=F)
# ggplot(top20_S9,  aes(x=gene, y=dataset, fill=ZS9) )+
#   geom_tile()+
#   scale_fill_gradient2("Z-score",low="white", high="red", na.value="white")+  
#   scale_y_discrete(expand=c(0,0)) +
#   ylab("")+
#   xlab("")+
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=6),
#         panel.border = element_rect(colour = "black", fill=NA))
# ggsave(file="heatmap_genomic_clustering_bygene_ZscoreS9_top20_sigfit_1Mb_min10percent_IgDenovo_N45711_ttest.pdf", width=7,height=2.5)
```


# of the top 20, how many are shared in other cell types (say, top 1% of F scores per type)?
```{r, message=FALSE, warning=FALSE}
# only genes with at least one mutation in at least one dataset
Nsum = tapply(alldatasets2$N, INDEX=alldatasets2$gene, FUN=sum)
nonzerogenes = names(Nsum)[Nsum>0] #22643- top 1% is ranks 22417-22643
alldatasetsNoZeros = subset(alldatasets2, gene %in% nonzerogenes)
rank_focal_list = list()
for (i in 1:6){
  focal = subset(alldatasetsNoZeros, dataset==mycancertypes[i] ) #& IgTF==FALSE
  focal2 = focal[sample(1:nrow(focal), size=nrow(focal)),]
  focal2$ZS9rank = -1*(rank(focal2$ZS9, ties.method = c("first") )-nrow(focal2)-1)
  focal2$ZSIgrank = -1*(rank(focal2$ZSIg, ties.method = c("first") )-nrow(focal2)-1)
  rank_focal_list[[i]] = focal2
}
myrankDF = data.frame(do.call(rbind, rank_focal_list))
Nsubset_top1p = c()
for (i in 1:length(top20_S9_samples)){
  Nsubset_top1p[i] = nrow(subset(myrankDF, gene==top20_S9_samples[i] & ZS9rank<244))
}
top20tab = data.frame(gene=top20_S9_samples, Nsubset_top1p=Nsubset_top1p)

top20bmem_S9_samples = subset(myrankDF, dataset=="Memory_B" & ZS9rank<=10)$gene
Nsubset_top1p = c()
for (i in 1:length(top20bmem_S9_samples)){
  Nsubset_top1p[i] = nrow(subset(myrankDF, gene==top20bmem_S9_samples[i] & ZS9rank<244))
}
top10bmemtab = data.frame(gene=top20bmem_S9_samples, Nsubset_top1p=Nsubset_top1p)
top20bmem_S9_samples = subset(myrankDF, dataset=="Memory_B" & ZS9rank<=100)$gene
Nsubset_top1p = c()
for (i in 1:length(top20bmem_S9_samples)){
  Nsubset_top1p[i] = nrow(subset(myrankDF, gene==top20bmem_S9_samples[i] & ZS9rank<244))
}
top100bmemtab = data.frame(gene=top20bmem_S9_samples, Nsubset_top1p=Nsubset_top1p)
top20bmem_S9_samples = subset(myrankDF, dataset=="Memory_B" & ZS9rank<=5)$gene
Nsubset_top1p = c()
for (i in 1:length(top20bmem_S9_samples)){
  Nsubset_top1p[i] = nrow(subset(myrankDF, gene==top20bmem_S9_samples[i] & ZS9rank<244))
}
top5bmemtab = data.frame(gene=top20bmem_S9_samples, Nsubset_top1p=Nsubset_top1p)
```


## Create table of top genes by SIg t-test p-value
```{r, message=FALSE, warning=FALSE}
topN=5
topSIgZ_focal_list = list()
topSIgZ_focal_list_info = list()
# exlude CLL unmutated 
for (i in 1:6){
  focal = subset(alldatasets2, dataset==mycancertypes[i] & IgTF==FALSE) #
  focal2 = focal[order(focal$p_SIg, decreasing=F), ][1:topN, ]
  topSIgZ_focal_list_info[[i]] = focal2
  #focal3 = focal2[focal2$ZSIg>6,]
  focal3 = focal2
  topSIgZ_focal_list[[i]] = focal3$gene
}
names(topSIgZ_focal_list) = mycancertypes[1:6]

mysamples = unique(unlist(topSIgZ_focal_list))
topSIgZ_focalDF = subset(alldatasets2, gene %in% unique(unlist(topSIgZ_focal_list)))
## there are two DHRSX genes- we are only interested in the one on the X chrom
topSIgZ_focalDF = subset(topSIgZ_focalDF, !(gene=="DHRSX" & chr=="Y") )
topSIgZ_focalDF$gene = factor(topSIgZ_focalDF$gene, levels=unique(unlist(topSIgZ_focal_list)))
topSIgZ_focalDF$dataset = revalue(topSIgZ_focalDF$dataset, c("Memory_B"= "Memory B", "CLL_mutated" ="CLL mutated", "DLBC_lymphoma" ="DLBC lymphoma"  ,"Follicular_lymphoma"="Follicular lymphoma", "Multiple_myeloma"="Multiple myeloma", "Burkitt_lymphoma"   ="Burkitt lymphoma","CLL_unmutated"   ="CLL unmutated" ))
topSIgZ_focalDF$dataset = factor(topSIgZ_focalDF$dataset, levels=rev(mycancertypes2))

#topSIgZ_focalDF = read.table( file="dataset_genomic_clustering_bygene_ZscoreSIg_top5eachNonIg_N45711.txt", header=T, stringsAsFactors = F)
ggplot(subset(topSIgZ_focalDF),  aes(x=gene, y=dataset, fill=-log10(p_SIg)) )+
  geom_tile()+
  scale_fill_gradientn("p (-log10)", na.value="white", guide = guide_colorbar(frame.colour = "black", ticks.colour="black"), limits=c(0,63), colours=c("white","red","darkred","black"))+  
  scale_y_discrete(expand=c(0,0)) +
  ylab("")+
  xlab("")+
  #theme(axis.text.x = element_text(angle = 45, vjust = 1.05, hjust=1, size=8),
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1.05, size=7),
        axis.text.y = element_text(size = 8),
        legend.text=element_text(size=8),
        legend.title=element_text(size=8),
        legend.key.size = unit(0.35, 'cm'),
        panel.border = element_rect(colour = "black", fill=NA))
ggsave(file="heatmap_genomic_clustering_bygene_ZscoreSIg_top5eachNonIg_sigfit_1Mb_min10percent_IgDenovo_N45711_ttest.pdf", width=4.75,height=1.9)
#write.table(topSIgZ_focalDF, file="dataset_genomic_clustering_bygene_ZscoreSIg_top5eachNonIg_N45711.txt", col.names = T, 
```


## Create table of top 20 genes by SIg t-test p-value
```{r, message=FALSE, warning=FALSE}
# nonIg = subset(alldatasets2, IgTF==FALSE)
# top20_SIg_samples = unique(nonIg[order(nonIg$ZSIg, decreasing=T),][,"gene"])[1:20]
# #top20_SIg_samples = unique(alldatasets2[order(alldatasets2$ZSIg, decreasing=T),][,"gene"])[1:20]
# top20_SIg = alldatasets2[alldatasets2$gene %in% top20_SIg_samples, ]
# top20_SIg$dataset = factor(top20_SIg$dataset, levels=rev(mycancertypes))
# 
# mysub = subset(top20_SIg, dataset=="Memory_B")
# mygenes_top20 = mysub[order(mysub$ZSIg, decreasing = T), "gene"]
# top20_SIg$gene = factor(top20_SIg$gene, levels=mysub[order(mysub$ZSIg, decreasing = T), "gene"])
# 
# ggplot(top20_SIg,  aes(x=gene, y=dataset, fill=ZSIg) )+
#   geom_tile()+
#   scale_fill_gradient2("Z-score",low="white", high="red", na.value="white")+  
#   scale_y_discrete(expand=c(0,0)) +
#   ylab("")+
#   xlab("")+
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=6),
#         panel.border = element_rect(colour = "black", fill=NA))
# ggsave(file="heatmap_genomic_clustering_bygene_ZscoreSIg_top20_sigfit_1Mb_min10percent_IgDenovo_N45711_ttest.pdf", width=7,height=2.5)
```

# of the top 20, how many are shared in other cell types (say, top 1% of F scores per type)?
```{r, message=FALSE, warning=FALSE}
# only genes with at least one mutation in at least one dataset
rank_focal_list = list()
for (i in 1:6){
  focal = subset(alldatasetsNoZeros, dataset==mycancertypes[i] & IgTF==FALSE)
  focal2 = focal[sample(1:nrow(focal), size=nrow(focal)),]
  focal2$ZS9rank = -1*(rank(focal2$ZS9, ties.method = c("first") )-nrow(focal2)-1)
  focal2$ZSIgrank = -1*(rank(focal2$ZSIg, ties.method = c("first") )-nrow(focal2)-1)
  rank_focal_list[[i]] = focal2
}
myrankDFnoIg = data.frame(do.call(rbind, rank_focal_list))

top20bmem_SIg_samples = subset(myrankDFnoIg, dataset=="Memory_B" & ZSIgrank<=10)$gene
Nsubset_top1p_Ig = c()
for (i in 1:length(top20bmem_SIg_samples)){
  Nsubset_top1p_Ig[i] = nrow(subset(myrankDF, gene==top20bmem_SIg_samples[i] & ZSIgrank<242))
}
top10Ig_bmemtab = data.frame(gene=top20bmem_SIg_samples, Nsubset_top1p=Nsubset_top1p_Ig)
top20bmem_SIg_samples = subset(myrankDFnoIg, dataset=="Memory_B" & ZSIgrank<=100)$gene
Nsubset_top1p_Ig = c()
for (i in 1:length(top20bmem_SIg_samples)){
  Nsubset_top1p_Ig[i] = nrow(subset(myrankDF, gene==top20bmem_SIg_samples[i] & ZSIgrank<242))
}
top100Ig_bmemtab = data.frame(gene=top20bmem_SIg_samples, Nsubset_top1p=Nsubset_top1p_Ig)
top20bmem_SIg_samples = subset(myrankDFnoIg, dataset=="Memory_B" & ZSIgrank<=5)$gene
Nsubset_top1p_Ig = c()
for (i in 1:length(top20bmem_SIg_samples)){
  Nsubset_top1p_Ig[i] = nrow(subset(myrankDF, gene==top20bmem_SIg_samples[i] & ZSIgrank<242))
}
top5Ig_bmemtab = data.frame(gene=top20bmem_SIg_samples, Nsubset_top1p=Nsubset_top1p_Ig)
```


# Candidate genes
```{r, message=FALSE, warning=FALSE}
## Candidate genes from Kasar2015 paper
genes_cAID_q01 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Kasar2015_AIDgenes/cAID_Kasar2015_genes_q01_gencodeID.csv", stringsAsFactors = FALSE,header=TRUE))
genes_ncAID_q01 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Kasar2015_AIDgenes/ncAID_Kasar2015_genes_q01_gencodeID.csv", stringsAsFactors = FALSE,header=TRUE))
## ALL represented, except one that I couldn't find: AC096579.7
#genes_cAID_q01$new_gene[!(genes_cAID_q01$new_gene %in% genann[,4])]
## ALL represented
#genes_ncAID_q01$new_gene[(genes_ncAID_q01$new_gene %in% genann[,4])]
subset(genann2, gene %in% genes_ncAID_q01$new_gene)
subset(genann2, gene %in% genes_cAID_q01$new_gene)

## Candidate SHM genes from Khodabakhshi2012 paper
genes_Khod2012 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Khodabakhshi2012_SHMgenes/Khodabakhshi2012_topgenes_Table1_gencodeID.csv", header=TRUE, stringsAsFactors = FALSE))
```

# OLD mem B analysis
```{r, message=FALSE, warning=FALSE}
 oldmemb = read.table(file=paste("../pergene_sigfit_1Mb_min10percent/S9cAIDblood_Z_per_gene_sigfit_Memory_B_1Mb_min10percent_N45844.txt", sep=""), header=T, stringsAsFactors = F)
```

