---
title: "sigfit_memoryB_1Mb_min10percent"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache=TRUE, fig.path="./graphics/plot",autodep=TRUE,fig.width=5, fig.height=4)
knitr::opts_knit$set(root.dir = '/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/sigfit/sigfit_memoryB_1Mb_min10percent')
library(ggplot2)
library(cowplot)
library(caTools)
library(reshape2)
library(dplyr)
library(hdp)
library(RColorBrewer)
library(grid)
library(gridExtra)
library(ggpubr)
#location="farm"
location="local"

if (location=="farm"){
  root="/lustre/scratch116/casm/cgp/users/hm8"
} else if (location=="local"){
  root="/Users/hm8/volumes/hm8_network"
} else {
  warning("location not set- using current directory")
  root=getwd()
}
```





###### load in data
```{r, message=FALSE, warning=FALSE}
groupname="memoryB_1Mb"

chromsize = read.table("/Users/hm8/sanger/genomes/human/GRCh37d5/hg19.chrom_sizes.txt", stringsAsFactors = FALSE)
chroms = chromsize[,1]

# per chrom
perchrom_lower_list = perchrom_upper_list = perchrom_mean_list = perchrom_info_list = list()
for (i in 1:length(chroms)){
  chrom=chroms[i]
  load(file=paste(root, "/lymphocyteWGS/mutsig_byregion/sigfit/sigfit_memoryB_1Mb_min10percent/exposures_", groupname, ".sigfit_union_hdp_sigprofiler_cosmic2.chr", chrom,".Rdata", sep=""))
  perchrom_lower_list[[i]] = exposures$lower_90
  perchrom_upper_list[[i]] = exposures$upper_90
  perchrom_mean_list[[i]] = exposures$mean
  perchrom_info_list[[i]] = samp_type
}
exposures_mean = data.frame(do.call(rbind, perchrom_mean_list))
exposures_lower = data.frame(do.call(rbind, perchrom_lower_list))
exposures_upper = data.frame(do.call(rbind, perchrom_upper_list))
samp_type = data.frame(do.call(rbind, perchrom_info_list))
sigprop = cbind(samp_type, exposures_mean)
```

###### load in sample size info
```{r, message=FALSE, warning=FALSE}
## read in data
# load("/Users/hm8/volumes/hm8_network/lymphocyte_somatic_mutations_final/mutsig_byregion/results/mutmats_byregion_bin1000000.Rdata")
# mysum = apply(mutmax, MARGIN=2, FUN=sum)
# mysample = paste(mutmaxinfo[,1], mutmaxinfo[,2], sep="_")
# mytotals = data.frame(samp_names=mysample, N=mysum)
# write.table(mytotals, file=paste(root, "/lymphocyteWGS/mutsig_byregion/sigfit/sigfit_memoryB_1Mb_min10percent/mutations_per_bin_", groupname, ".txt", sep=""), col.names = T, row.names = F, sep="\t", quote=F)
myN = read.table(file=paste(root, "/lymphocyteWGS/mutsig_byregion/sigfit/sigfit_memoryB_1Mb_min10percent/mutations_per_bin_", groupname, ".txt", sep=""), header=T, stringsAsFactors = F)
```


###### label Ig regions
```{r, message=FALSE, warning=FALSE}
  # identify Ig regions (+/1 100KB)
# if either of the ends lies inside the following Ig OR
# if either of the Ig coordinates lies inside the bins
ighst = 106304735-100000 # 14
ighend = 107283226+100000 # 14
iglst = 22385390-100000 # 22
iglend = 23263607+100000 # 22
igkst = 89160078-100000 # 2
igkend = 90274237+100000 # 2
cs_start = 106053274-100000 #14
cs_end = 106322322+100000 #14
annotate_vdj_vector = function(X2){
  X = X2
  X[1] = as.character(X2[1])
  X[2] = as.numeric(X2[2])
  X[3] = as.numeric(X2[3])
  if ( X[1] == "14" & ( (X[2]>=ighst & X[2]<=ighend) | (X[3]>=ighst & X[3]<=ighend) ) ) {return("igh")} else
    if ( X[1] == "22" & ( (X[2]>=iglst & X[2]<=iglend) | (X[3]>=iglst & X[3]<=iglend) ) ) {return("igl")} else
      if ( X[1] == "2" & ( (X[2]>=igkst & X[2]<=igkend) | (X[3]>=igkst & X[3]<=igkend) ) ) {return("igk")} else
        if ( X[1] == "14" & ( (X[2]>=cs_start & X[2]<=cs_end) | (X[3]>=cs_start & X[3]<=cs_end) ) ) {return("classS")} else
   if ( X[1] == "14" & ( (X[2]>=ighst & X[3]<=ighst) | (X[2]>=ighend & X[3]<=ighend) ) ) {return("igh")} else
    if ( X[1] == "22" & ( (X[2]>=iglst & X[3]<=iglst) | (X[2]>=iglend & X[3]<=iglend) ) ) {return("igl")} else
      if ( X[1] == "2" & ( (X[2]>=igkst & X[3]<=igkst) | (X[2]>=igkend & X[3]<=igkend) ) ) {return("igk")} else
        if ( X[1] == "14" & ( (X[2]>=cs_start & X[3]<=cs_start) | (X[2]>=cs_end & X[3]<=cs_end) ) ) {return("classS")} else {return(FALSE)}
}
myVDJ = apply(data.frame(sigprop$chr, sigprop$start, sigprop$end), MARGIN=1, annotate_vdj_vector)
sigprop$Ig = myVDJ
sigprop$IgTF = TRUE
sigprop$IgTF[sigprop$Ig==FALSE] = FALSE
sig_prop_type = merge(myN, sigprop[c(ncol(sigprop)-1, ncol(sigprop), 1:(ncol(sigprop)-2) )],  by="samp_names")
sig_N_type = sig_prop_type
sig_N_type[,8:ncol(sig_prop_type)] = sig_prop_type[,8:ncol(sig_prop_type)]*sig_prop_type$N
```


### plotting Nmutations vs prop S9
```{r, message=FALSE, warning=FALSE}
s9_2sd_mean = 2*sd(sig_N_type$SBS9, na.rm=TRUE) + mean(sig_N_type$SBS9, na.rm=TRUE)
sig_N_type_2sd_SBS9 = na.omit(sig_N_type[sig_N_type$SBS9 > s9_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig","IgTF")])
s31_2sd_mean = 2*sd(sig_N_type$Signature.Ig, na.rm=TRUE) + mean(sig_N_type$Signature.Ig, na.rm=TRUE)
sig_N_type_2sd_Signature.Ig = na.omit(sig_N_type[sig_N_type$Signature.Ig > s31_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig","IgTF")])
Signature.blood_2sd_mean = 2*sd(sig_N_type$Signature.blood, na.rm=TRUE) + mean(sig_N_type$Signature.blood, na.rm=TRUE)
sig_N_type_2sd_Signature.blood = na.omit(sig_N_type[sig_N_type$Signature.blood > Signature.blood_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig","IgTF")])

sig_N_type_2sd_SBS9$sig9 = TRUE
sig_N_type_2sd_Signature.Ig$sigIg = TRUE

tmp1 = merge(sig_prop_type, sig_N_type_2sd_Signature.Ig[,c("chr","start","sigIg")], by=c("chr","start"), all.x=TRUE)
sig_prop_type_sigs = merge(tmp1, sig_N_type_2sd_SBS9[,c("chr","start","sig9")], by=c("chr","start"), all.x=TRUE)
sig_prop_type_sigs$mysigs = c("not significant")
sig_prop_type_sigs$mysigs[sig_prop_type_sigs$sigIg==TRUE] = c("cAID")
sig_prop_type_sigs$mysigs[sig_prop_type_sigs$sig9==TRUE] = c("ncAID")

ggplot(subset(sig_prop_type_sigs), aes(x=N, SBS9, color=mysigs, alpha=mysigs, shape=IgTF))+
  geom_point()+
  scale_color_manual("", values=c("cyan","#984EA3","black"), labels=c("cAID","ncAID",""))+
  scale_alpha_manual("", values=c(1,0.3,0.1), labels=c("cAID","ncAID",""))+
  xlab("# mutations per 1MB bin")+
  ylab("Proportion of S9 mutations")+
  theme_light()
#ggsave(file="figures/nmut1MBbins_propS9_scatter.pdf", width=5, height=2.5)


ggplot(subset(sig_prop_type_sigs, Ig == FALSE), aes(x=N, SBS9, color=mysigs, alpha=mysigs))+
  geom_point()+
  scale_color_manual("", values=c("cyan","#984EA3","black"), labels=c("cAID","ncAID",""))+
  scale_alpha_manual("", values=c(1,0.3,0.1), labels=c("cAID","ncAID",""))+
  xlab("# mutations per 1MB bin")+
  ylab("Proportion of S9 mutations")+
  theme_light()
#ggsave(file="figures/nmut1MBbins_propS9_noIg_scatter.pdf", width=5, height=2.5)


```


## how does this compare with the hdp results?
```{r, message=FALSE, warning=FALSE}
name_sigs="CosmicIG"
my_prior=1000
groupname="by_1MB"
load(file=paste("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/sig_prop_.", name_sigs, ".", my_prior,".",groupname, ".Rdata", sep="") )
head(sig_prop_type)
plot(sig_prop_type)

## loads over my objects, 
samp_type = data.frame(do.call(rbind, perchrom_info_list))
sigprop = cbind(samp_type, exposures_mean)

sig_prop_type_both = merge(sigprop, sig_prop_type, by=1)
ggplot(sig_prop_type_both, aes(SBS9, SBS9))+
  geom_point(alpha=0.5)+
  xlab("sigfit S9 proportion / 1Mb")+
  ylab("hdp S9 proportion / 1Mb")+
  theme_light()
ggsave(paste("figures/S9prop_sigfit_vs_hdp.", groupname, ".pdf", sep=""), width=4,height=3)
```

## how does this compare with the hdp results?
```{r, message=FALSE, warning=FALSE}
ggplot(sig_prop_type_both, aes(SBS9, Signature.Ig))+
  geom_point(alpha=0.5)+
  xlab("sigfit S9 proportion / 1Mb")+
  ylab("sigfit cAID proportion / 1Mb")+
  theme_light()
ggsave(paste("figures/S9prop_vs_cAIDprop.", groupname, ".pdf", sep=""), width=4,height=3)
```


## how does this compare with the hdp results?
```{r, message=FALSE, warning=FALSE}
ggplot(sig_prop_type_both, aes(Signature.Ig, Signature.Ig))+
  geom_point(alpha=0.5)+
  xlab("sigfit cAID proportion / 1Mb")+
  ylab("hdp cAID proportion / 1Mb")+
  theme_light()
ggsave(paste("figures/cAIDprop_sigfit_vs_hdp.", groupname, ".pdf", sep=""), width=4,height=3)
```


## Distribution of ncAID and cAID across genome
```{r, message=FALSE, warning=FALSE}
my10 = brewer.pal(9, "Set1") 
my_colors = c("grey",my10[c(4)],"cyan")
myregions=data.frame(x1=c(106304735,22385390), x2=c(107283226,23263607), y1=c(0,0), y2=c(400,400), chr=c('14','22'), region=c("IGH","IGL"))

myline=data.frame(x1=c(106304735,22385390), x2=c(107283226,23263607), y1=c(0,0), y2=c(400,400), chr=c('14','22'), region=c("IGH","IGL"))

# s9_2sd_mean = 2*sd(sig_N_type$SBS9, na.rm=TRUE) + mean(sig_N_type$SBS9, na.rm=TRUE)
# sig_N_type_2sd_SBS9 = na.omit(sig_N_type[sig_N_type$SBS9 > s9_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig")])
# s31_2sd_mean = 2*sd(sig_N_type$Signature.Ig, na.rm=TRUE) + mean(sig_N_type$Signature.Ig, na.rm=TRUE)
# sig_N_type_2sd_Signature.Ig = na.omit(sig_N_type[sig_N_type$Signature.Ig > s31_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig")])
# Signature.blood_2sd_mean = 2*sd(sig_N_type$Signature.blood, na.rm=TRUE) + mean(sig_N_type$Signature.blood, na.rm=TRUE)
# sig_N_type_2sd_Signature.blood = na.omit(sig_N_type[sig_N_type$Signature.blood > Signature.blood_2sd_mean ,c("chr","start","Signature.blood","SBS9","Signature.Ig")])

newvec_list = list()
for (i in 1:length(unique(sig_N_type$chr))){
  focal = subset(sig_N_type, chr==unique(sig_N_type$chr)[i])
  newvec_list[[i]] = order(focal$start)
}
newvec = unlist(newvec_list)
sig_N_type$start_rank = newvec
sig_N_type_melt = melt(sig_N_type[,c("chr","start","start_rank","IgTF","Signature.blood","SBS9","Signature.Ig")], id.vars=c("chr","start","start_rank","IgTF"))
sig_N_type_melt$valuepergenome = sig_N_type_melt$value/74
## 74 samples

sig_N_type_melt_auto = subset(sig_N_type_melt, chr %in% 1:22)
sig_N_type_melt_sex =  subset(sig_N_type_melt, chr %in% c("X","Y"))
sig_N_type_melt_auto = sig_N_type_melt_auto[order(as.numeric(sig_N_type_melt_auto$chr), as.numeric(sig_N_type_melt_auto$start)),]
sig_N_type_melt_auto$chr = factor(sig_N_type_melt_auto$chr, levels=1:22)
sig_N_type_melt2 = rbind(sig_N_type_melt_auto, sig_N_type_melt_sex)

# mean(subset(sig_N_type, sig_N_type$Signature.Ig<=s31_2sd_mean)$Signature.Ig)/65
# mean(subset(sig_N_type, sig_N_type$Signature.Ig>s31_2sd_mean)$Signature.Ig)/65
# mean(subset(sig_N_type, sig_N_type$SBS9<=s9_2sd_mean)$SBS9)/65
# mean(subset(sig_N_type, sig_N_type$SBS9>s9_2sd_mean)$SBS9)/65
# median(sig_N_type$Signature.Ig, na.rm=TRUE)/65
# median(sig_N_type$SBS9, na.rm=TRUE)/65
# median(sig_N_type$Signature.blood, na.rm=TRUE)/65
 

ggplot(sig_N_type_melt2)+
  geom_line(data = myregions, aes(x = x1, xend = x2, y = 0, yend = Inf), alpha = 1, fill = 'black', lty=2) +
  geom_col(aes(start, value, fill=variable), position=position_dodge(), alpha=1, width=5000000)+
  geom_point(data=sig_N_type_2sd_SBS9, aes(start+500000, SBS9), col=my10[c(4)], alpha=1, pch=1, size=1)+
  geom_point(data=sig_N_type_2sd_Signature.Ig, aes(start+1100000, Signature.Ig), col="cyan", alpha=1, pch=1, size=1)+
  facet_grid(.~chr, scales="free_x", space = "free_x")+
  scale_fill_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  scale_color_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  xlab("Genomic position")+
  ylab("# mutations")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,350)) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.spacing = unit(0, "lines"),
        axis.line = element_line(colour = "black"),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
  # theme(panel.grid.major = element_blank(), 
  #      panel.grid.minor = element_blank(),
  #   axis.title.y.right = element_blank(),                # hide right axis title
  #       axis.text.y.right = element_blank(),                 # hide right axis labels
  #       axis.ticks.y = element_blank(),                      # hide left/right axis ticks
  #       axis.text.y = element_text(margin = margin(r = 0)),  # move left axis labels closer to axis
  #       panel.spacing = unit(0, "mm"),                       # remove spacing between facets
  #       strip.background = element_rect(size = 0.5)) 
#ggsave(file="figures/nAID_1MB_genomic_barplot_spacedperbin_sigfit_Sept2020.pdf", width=10, height=3)

#
ggplot(sig_N_type_melt2)+
  geom_line(data = myregions, aes(x = x1, xend = x2, y = 0, yend = Inf), alpha = 1, fill = 'black', lty=2) +
  geom_col(aes(start, value/74, fill=variable), position=position_dodge(), alpha=1, width=5000000)+
  geom_point(data=sig_N_type_2sd_SBS9, aes(start+500000, SBS9/74), col=my10[c(4)], alpha=1, pch=1, size=1)+
  geom_point(data=sig_N_type_2sd_Signature.Ig, aes(start+1100000, Signature.Ig/74), col="cyan", alpha=1, pch=1, size=1)+
  facet_grid(.~chr, scales="free_x", space = "free_x")+
  scale_fill_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  scale_color_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  xlab("Genomic position")+
  ylab("# mutations")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,5)) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.spacing = unit(0, "lines"),
        axis.line = element_line(colour = "black"),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
#ggsave(file="figures/nAID_1MB_genomic_barplot_spacedperbin_sigfit_Sept2020_Nper74ind.pdf", width=10, height=3)



######### No Ig
ggplot(subset(sig_N_type_melt2, IgTF==FALSE) )+
  #geom_line(data = myregions, aes(x = x1, xend = x2, y = 0, yend = Inf), alpha = 1, fill = 'black', lty=2) +
  geom_col(aes(start, value, fill=variable), position=position_dodge(), alpha=1, width=5000000)+
  geom_point(data=subset(sig_N_type_2sd_SBS9, IgTF==FALSE), aes(start+500000, SBS9), col=my10[c(4)], alpha=1, pch=1, size=1)+
  geom_point(data=subset(sig_N_type_2sd_Signature.Ig, IgTF==FALSE), aes(start+1100000, Signature.Ig), col="cyan", alpha=1, pch=1, size=1)+
  facet_grid(.~chr, scales="free_x", space = "free_x")+
  scale_fill_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  scale_color_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  xlab("Genomic position")+
  ylab("# mutations")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,120)) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.spacing = unit(0, "lines"),
        axis.line = element_line(colour = "black"),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
#ggsave(file="figures/nAID_1MB_genomic_barplot_spacedperbin_sigfit_Sept2020_noIg.pdf", width=10, height=3)

#
ggplot(subset(sig_N_type_melt2, IgTF==FALSE) )+
  geom_line(data = myregions, aes(x = x1, xend = x2, y = 0, yend = Inf), alpha = 1, fill = 'black', lty=2) +
  geom_col(aes(start, value/74, fill=variable), position=position_dodge(), alpha=1, width=5000000)+
  geom_point(data=subset(sig_N_type_2sd_SBS9, IgTF==FALSE) , aes(start+500000, SBS9/74), col=my10[c(4)], alpha=1, pch=1, size=1)+
  geom_point(data=subset(sig_N_type_2sd_Signature.Ig, IgTF==FALSE), aes(start+1100000, Signature.Ig/74), col="cyan", alpha=1, pch=1, size=1)+
  facet_grid(.~chr, scales="free_x", space = "free_x")+
  scale_fill_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  scale_color_manual("Mutation type", labels=c("Sblood","ncAID (S9)","cAID"), values=my_colors)+
  xlab("Genomic position")+
  ylab("# mutations")+
  scale_y_continuous(expand = c(0, 0), limits=c(0,1.65)) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.spacing = unit(0, "lines"),
        axis.line = element_line(colour = "black"),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
#ggsave(file="figures/nAID_1MB_genomic_barplot_spacedperbin_sigfit_Sept2020_Nper74ind_noIg.pdf", width=10, height=3)
```



## Closest gene(s) to 1MB bin
```{r, message=FALSE, warning=FALSE}
genann = read.table("/Users/hm8/sanger/genomes/human/GRCh37d5/gencode/gencode.v30lift37.basic.annotation.geneonly.genename.bed", stringsAsFactors = F)

genelist = list()
genetable = sig_prop_type_sigs
genetable$genes = NA
genetable$Ngenes = 0
for (i in 1:nrow(sig_prop_type_sigs)){  #  3113 1MB bins
#  for (i in 1:10){  #  3113 1MB bins

  focal = sig_prop_type_sigs[i,c("chr","start","end")]
  #geneoverlap = genann[ as.character(genann[,1])==as.character(focal[1]) & 
  #      ( ( as.numeric(genann[,2])>=as.numeric(focal[2]) & as.numeric(genann[,2])<as.numeric(focal[3])) | (as.numeric(genann[,3])>=as.numeric(focal[2]) & as.numeric(genann[,3])<as.numeric(focal[3])) ), ] ## Puts genes overlapping bin into both
  focal_chrom = as.character(focal$chr)
  genann_focal = subset(genann, V1 == focal_chrom)
  geneoverlap = genann_focal[  (as.numeric(genann_focal[,2])>=as.numeric(focal[2]) &   # gene start within bin
                                  as.numeric(genann_focal[,2])<=as.numeric(focal[3]) ) 
                              | (as.numeric(genann_focal[,3])>=as.numeric(focal[2]) &  # gene end within bin
                                  as.numeric(genann_focal[,3])<=as.numeric(focal[3]) )
                              | (as.numeric(genann_focal[,2])<as.numeric(focal[2]) &
                                   as.numeric(genann_focal[,3])>as.numeric(focal[3]) )  
                              , ] # gene start and end surround bin
                      ## Puts genes overlapping bins into BOTH bins
                      
  genelist[[i]] = unique(geneoverlap[,4])
  genetable$genes[i] = paste(unique(geneoverlap[,4]), collapse=";")
  genetable$Ngenes[i] = length(unique(geneoverlap[,4]))
}
genetable$N_Signature.Ig = genetable$Signature.Ig*genetable$N
genetable$N_SBS9 = genetable$SBS9*genetable$N
write.table(genetable[order(genetable$N_Signature.Ig, decreasing=T),], file="sig_prop_type_sigs_genes.txt", col.names=T, row.names=F, quote=F, sep="\t")
write.table(subset(genetable,mysigs=="cAID" & IgTF == FALSE) , file="sig_prop_type_sigs_genes_4cAID.txt", col.names=T, row.names=F, quote=F, sep="\t")
allgenes = unlist(genelist)
length(allgenes) # 59379 genes (where )
length(unique(allgenes)) # 57688
notincluded = genann[!(genann[,4] %in% unique(allgenes) ),] # 99
```




# Candidate genes
```{r, message=FALSE, warning=FALSE}
## Candidate genes from Kasar2015 paper
genes_cAID_q01 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Kasar2015_AIDgenes/cAID_Kasar2015_genes_q01_gencodeID.csv", stringsAsFactors = FALSE,header=TRUE))
genes_ncAID_q01 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Kasar2015_AIDgenes/ncAID_Kasar2015_genes_q01_gencodeID.csv", stringsAsFactors = FALSE,header=TRUE))
## ALL represented, except one that I couldn't find: AC096579.7
#genes_cAID_q01$new_gene[!(genes_cAID_q01$new_gene %in% genann[,4])]
## ALL represented
#genes_ncAID_q01$new_gene[(genes_ncAID_q01$new_gene %in% genann[,4])]
subset(genann2, gene %in% genes_ncAID_q01$new_gene)
subset(genann2, gene %in% genes_cAID_q01$new_gene)

## Candidate SHM genes from Khodabakhshi2012 paper
genes_Khod2012 = na.omit(read.csv("/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/Khodabakhshi2012_SHMgenes/Khodabakhshi2012_topgenes_Table1_gencodeID.csv", header=TRUE, stringsAsFactors = FALSE))
```

# Do the 4 cAID regions overlap with top cAID genes in the Kasar2015 paper?
```{r, message=FALSE, warning=FALSE}
## Candidate genes from Kasar2015 paper
cAID_top4_genes = unlist(lapply(subset(genetable, mysigs == "cAID" & IgTF == FALSE)$genes, FUN=function(X) unlist(strsplit(X, split=";")) ) )
genes_cAID_q01[genes_cAID_q01$gene %in% cAID_top4_genes,]
# only BCL6

merge_Khod = subset(genes_Khod2012, IGH.NON_IGH=="NON_IGH" & new_gene %in% cAID_top4_genes)
# of the 43 top genes, 4 are in our top 4 regions (represents 96 genes)
merge_Khod$new_gene
# [1] "BCL6"  "BCL7A" "PAX5"  "GRHPR"
```

## Per-gene mutations
```{r, message=FALSE, warning=FALSE}
## read in per locations
genann = read.table("/Users/hm8/sanger/genomes/human/GRCh37d5/gencode/gencode.v30lift37.basic.annotation.geneonly.genename.bed", stringsAsFactors = F)
## Also has the following extra chromosomes not in our analysis
# [25] "M"          "GL000193.1" "GL000192.1" "GL000237.1" "GL000220.1" "GL000241.1"
# [31] "GL000199.1" "GL000212.1" "GL000202.1" "GL000228.1" "GL000195.1" "GL000205.1"
# [37] "GL000204.1"
### Hugo Looked: not used
# source("/Users/hm8/sanger/genes/hugo_genes/hugoLookup3_DanL.R")
# genann_hugo = hugoLookup(genann[,4]) 
# genann[,5] = genann_hugo
# colnames(genann) = c("chr","start","end","gencode","hugo")
# save(genann, file="gencode.v30lift37.basic.annotation.geneonly.genename.hugo.bed")


## read in per bp annotation
myatt = read.table(paste(root, "/lymphocyteWGS/mutsig_byregion/sigfit/sigfit_memoryB_1Mb_min10percent/attribution_perbp_sigfit_memoryB_1Mb_min10percent.txt", sep=""), header=T, stringsAsFactors = F, sep="\t")
myatt = myatt[sample(1:nrow(myatt), size=45844),] # down sample to match cancer genomes
outlist = list()
for (i in 1:nrow(genann)){
  # add per-gene info: Nmutations, NS9, meanS9,  NSIg, meanSIg, NSblood, meanSSblood
  focal = unlist(genann[i,])
  focal_ann = subset(myatt, chr==focal[1] & pos >= as.numeric(focal[2]) & pos <= as.numeric(focal[3]) )
  if (nrow(focal_ann)==0){
    outlist[[i]] = unlist(c(focal, rep(0, times=7)))
  } else {
    outlist[[i]] = unlist(c(focal, nrow(focal_ann), sum(focal_ann$SBS9), mean(focal_ann$SBS9), sum(focal_ann$Signature.Ig), mean(focal_ann$Signature.Ig), sum(focal_ann$Signature.blood), mean(focal_ann$Signature.blood) ))
  }
}
genann2 = data.frame(do.call(rbind, outlist))
colnames(genann2) = c("chr","start","end","gene","N","NS9", "meanS9",  "NSIg", "meanSIg", "NSblood", "meanSblood")
genann2$N = as.numeric(as.character(genann2$N))
genann2$NS9 = as.numeric(as.character(genann2$NS9))
genann2$meanS9 = as.numeric(as.character(genann2$meanS9))
genann2$NSIg = as.numeric(as.character(genann2$NSIg))
genann2$meanSIg = as.numeric(as.character(genann2$meanSIg))
genann2$NSblood = as.numeric(as.character(genann2$NSblood))
genann2$meanSblood = as.numeric(as.character(genann2$meanSblood))
myVDJ = apply(data.frame(genann2$chr, genann2$start, genann2$end), MARGIN=1, annotate_vdj_vector)
genann2$VDJ = myVDJ

## Z statistic- deviation from mean
## magnitude 
Z_S9 =  Z_SIg =  Z_Sblood = c()
meanS9 = mean(myatt$SBS9, na.rm=TRUE)
meanSIg = mean(myatt$Signature.Ig, na.rm=TRUE)
meanSblood = mean(myatt$Signature.blood, na.rm=TRUE)
for (i in 1:nrow(genann2)){
  focal = genann2[i,]
  if(focal$N == 0 ){
    Z_S9[i] =  Z_SIg[i] =  Z_Sblood[i] = 0
  } else {
   Z_S9[i] = (focal$meanS9 - meanS9) / (sqrt(focal$meanS9*(1-focal$meanS9)/focal$N)) 
   Z_SIg[i] = (focal$meanSIg - meanSIg) / (sqrt(focal$meanSIg*(1-focal$meanSIg)/focal$N)) 
  Z_Sblood[i] = (focal$meanSblood - meanSblood) / (sqrt(focal$meanSblood*(1-focal$meanSblood)/focal$N))
  }
}
genann2$ZS9 = Z_S9
genann2$ZSIg = Z_SIg
genann2$ZSblood = Z_Sblood
# par(mfrow=c(2,2))
# hist(subset(genann2, N>0)$ZS9)
# hist(genann2$ZSIg)
# hist(genann2$ZSblood)
write.table(genann2, file="/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/mutsig_byregion/genomic_distribution_vs_pcawg/pergene_sigfit_1Mb_min10percent/S9cAIDblood_Z_per_gene_sigfit_Memory_B_1Mb_min10percent_N45844.txt", col.names = T, row.names = F, quote=F, sep="\t")
```







