---
title: "KX002hsc_lymph SNV first filter July, 2019"
output: html_document
---
<style type="text/css">

body{ /* Normal  */
      font-size: 10px;
  }
td {  /* Table  */
  font-size: 8px;
}
h1.title {
  font-size: 24px;
  color: DarkRed;
}
h1 { /* Header 1 */
  font-size: 16px;
  color: DarkBlue;
}
h2 { /* Header 2 */
    font-size: 16px;
  color: DarkBlue;
}
h3 { /* Header 3 */
  font-size: 12px;
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 10px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 10px;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache=TRUE, fig.path="./graphics/plot",autodep=TRUE,fig.width=5, fig.height=4)
knitr::opts_knit$set(root.dir = '/Users/hm8/sanger/lymphocyteExpansionSequenceAnalysis/KX002hsc_lymph')
library(ggplot2)
library(cowplot)
library(caTools)
library(reshape2)
library(dplyr)
library(diptest)
library(quantmod)
```

## Reading in raw data and shearwater results
```{r, fig.width=2.5, fig.height=2}
groupname = "KX002hsc_lymph"
system("mkdir -p figures")
shearw = read.table(paste(groupname,".shearwater.vaf40.txt", sep=""), sep="\t", stringsAsFactors = F, header=T)
## Already filtered for mean vaf 40 (reduces file size significantly)
shearw$ID = paste(shearw$chr, ":", shearw$pos, "_", shearw$ref, "/", shearw$mut, sep="")  
colony = unlist(lapply(shearw$sampleID, FUN=function(X) unlist(strsplit(X, split=".", fixed=TRUE))[1]))
shearw$colony = colony   
length(unique(shearw$colony)) # number of colonies

## Original VAFs (mutation ID as rowname)
HQdp1 = read.table(paste(groupname,"_totalDP.txt", sep=""), stringsAsFactors = F, header=T, check.names=F)
HQalt1 = read.table(paste(groupname,"_alternateDP.txt", sep=""), stringsAsFactors = F, header=T, check.names=F)
ncol(HQdp1)

sample_names = colnames(HQdp1)
sum(sample_names %in% unique(shearw$colony) )  ## Are all represented in shearwater?
Nsamples = length(sample_names)  ##

HQdpNA = HQdp1
HQdpNA[HQdp1==0] = NA
HQfreq1 = HQalt1/HQdpNA
```
## Distribution of shearwater pvalues (post-VAF40% exclusion)
```{r}
## Once the germline variants are removed, the log-pvalues are double-peaked 
## For this dataset, is it bimodal? (low p-value)
shearwA40 = shearw
dip.test(log(shearwA40$pval))$p.value
dens1 = density(-log(shearwA40$pval))
dens1df = data.frame(xval = as.numeric(dens1$x), yval=as.numeric(dens1$y))

peak1 = findPeaks(dens1$y)
val1 = findValleys(dens1$y)
my_logval = dens1$x[val1][which(val1> peak1[1] & val1 <  peak1[2])]
{plot(dens1)
abline(v=dens1$x[val1], col="black")
abline(v=my_logval, col="red")}
my_valley = exp(-my_logval) 
my_valley
shearwA_pval_keep = unique(shearwA40[shearwA40$pval<my_valley, ]$ID)
length(shearwA_pval_keep)/length(unique(shearwA40$ID))  # retains what fraction of the sites?

```

## Filter out sites based on shearwater and colonies with less than 6x dp
```{r}
meandp = apply(HQdp1, MARGIN=2, FUN=mean, na.rm=TRUE)
meandp[which(meandp<10)]  ## which have a mean coverage less than 10?
dpkeep = which(meandp>=6) & sample_names != "PD40667al"
sample_names[meandp<6]
sample_names = sample_names[dpkeep]

hist(meandp, main="Mean dp per colony", breaks=20, xlab="dp")
HQdp = subset(HQdp1, rownames(HQdp1) %in% shearwA_pval_keep)[,dpkeep]
HQalt = subset(HQalt1, rownames(HQalt1) %in% shearwA_pval_keep)[,dpkeep]
HQfreq =subset(HQfreq1, rownames(HQfreq1) %in% shearwA_pval_keep)[,dpkeep]
```

## Using the overdispersion filter
```{r}
## Identify somatic variants
######### The dispersion filter
##Note: Overdispersion function currently cannot take NAs
bb_df = na.omit(read.table(paste(groupname,"_rho.txt", sep=""), stringsAsFactors = FALSE, header=TRUE))
rho_est = as.numeric(bb_df$rho)
pvalue = as.numeric(bb_df$pvalue)

dip.test(log(rho_est))$p.value # 0
dens1 = density(-log(rho_est))
dens1df = data.frame(xval = as.numeric(dens1$x), yval=as.numeric(dens1$y))

peak1 = findPeaks(dens1$y)
val1 = findValleys(dens1$y)
my_logval = dens1$x[val1][1]

{plot(dens1)
abline(v=dens1$x[val1], col="black")
abline(v=my_logval, col="red")}

my_valley2 = my_logval
my_valley2
```

## Setting threshold
```{r, message=FALSE, warning=FALSE}
log_rho_threshold = -my_valley2
pvalue_threshold = 0.1
{plot(log(rho_est),pvalue, col=rgb(0,0,0,0.5))
  abline(h=pvalue_threshold, col="blue")
  abline(v=log_rho_threshold, col="red")}

```

## Filtering based on overdispersion test
```{r, message=FALSE, warning=FALSE, fig.width=2, fig.height=2}
flt_rho_toss = bb_df$ID[log(rho_est)<=log_rho_threshold|pvalue>=pvalue_threshold]
flt_rho_keep = bb_df$ID[log(rho_est)>log_rho_threshold & pvalue<pvalue_threshold] 
length(flt_rho_keep)/(length(flt_rho_keep)+length(flt_rho_toss)) # Proportion to keep
HQalt_keep = subset(HQalt, !(rownames(HQalt) %in% flt_rho_toss) )
HQfreq_keep = subset(HQfreq, !(rownames(HQfreq) %in% flt_rho_toss) )
```


## Identifying low frequency sites (potenitally invitro mutations) and min alt < 2
```{r, message=FALSE, warning=FALSE}
tmp = data.frame(HQfreq_keep, HQalt_keep)
Nsamples = ncol(HQfreq_keep)

meanVAF_alt2_keep = apply(tmp, MARGIN=1, FUN=function(X) mean( X[1:Nsamples][X[(Nsamples+1):(2*Nsamples)]>=2] ) )
meanVAF_alt2_keep[is.nan(meanVAF_alt2_keep)] = 0
Nsamples_alt2_keep = apply(tmp, MARGIN=1, FUN=function(X) length( X[1:Nsamples][ X[(Nsamples+1):(2*Nsamples)]>=2 ] ) )

meanVAF_alt2_freq20_keep = apply(tmp, MARGIN=1, FUN=function(X) mean( X[1:Nsamples][    X[1:Nsamples]>0.2 & X[(Nsamples+1):(2*Nsamples)]>=2 ] ) )
meanVAF_alt2_freq20_keep[is.nan(meanVAF_alt2_freq20_keep)] = 0
Nsamples_alt2_freq20_keep = apply(tmp, MARGIN=1, FUN=function(X) length( X[1:Nsamples][    X[1:Nsamples]>0.2 & X[(Nsamples+1):(2*Nsamples)]>=2 ] ) )

HQfreq_keep$meanVAF_alt2 = as.numeric(meanVAF_alt2_keep)
HQfreq_keep$Nsamples_alt2 = Nsamples_alt2_keep

HQfreq_keep$meanVAF_alt2_freq20 = as.numeric(meanVAF_alt2_freq20_keep)
HQfreq_keep$Nsamples_alt2_freq20 = Nsamples_alt2_freq20_keep
```

## Plotting effect of filter
```{r, message=FALSE, warning=FALSE}
{hist( as.numeric(HQfreq_keep$meanVAF_alt2[HQfreq_keep$Nsamples_alt2>1]), breaks=100, freq=F, main="Mean VAF (alt2), keep", xlim=c(0,1), xlab="VAF")
hist( as.numeric(HQfreq_keep$meanVAF_alt2[HQfreq_keep$Nsamples_alt2==1]), breaks=100, freq=F, col=rgb(1,0,0,0.5), add=TRUE)
legend("topright", fill=c("white", rgb(1,0,0,0.5)), legend=c("Shared","Private"))}

{hist( as.numeric(HQfreq_keep$meanVAF_alt2_freq20[HQfreq_keep$Nsamples_alt2_freq20>1]), breaks=100, freq=F, main="Mean VAF (alt2 freq20), keep", xlim=c(0,1), xlab="VAF")
hist( as.numeric(HQfreq_keep$meanVAF_alt2_freq20[HQfreq_keep$Nsamples_alt2_freq20==1]), breaks=100, freq=F, col=rgb(1,0,0,0.5), add=TRUE)
legend("topright", fill=c("white", rgb(1,0,0,0.5)), legend=c("Shared","Private"))}
```

## Shared mutations among colonies 
```{r}
t1 = table(HQfreq_keep$Nsamples_alt2_freq20)
t0 = data.frame(Ncolonies=as.numeric(names(t1)[names(t1)>0]), Nsnvs = t1[names(t1)>0])
ggplot(t0)+
  geom_point(aes(as.numeric(Ncolonies), Nsnvs.Freq))+
  scale_y_continuous(trans='log10') + 
  ylab("# SNVs (log)") + 
  xlab("# colonies sharing SNV") 
#ggsave(file=paste("figures/colonysharing_", groupname,".pdf", sep=""), width=6,height=4)
```


## Filtering out SNVs identified based on low VAF or alt depth
```{r, message=FALSE, warning=FALSE}
## Assigning NAs to anything below cutoff
HQfreqsomNA = HQfreq_keep[,1:Nsamples]
HQfreqsomNA[ (tmp[, 1:Nsamples]<=0.2 | tmp[, (Nsamples+1):(2*Nsamples)]<2) ] = NA  ## Anytime the alt is <2 or freq <0.2
##HQfreqsomNA[ (tmp[, (Nsamples+1):(2*Nsamples)]<2) ] = NA 
HQfreqsomNA2 = merge( unique(shearw[ , c("ID", "chr", "pos", "ref", "mut")]), data.frame(rownames(HQfreqsomNA), HQfreqsomNA), by=1)
```


## Checking VAF distribution of filtered data
```{r, message=FALSE, warning=FALSE, fig.width=2, fig.height=2}
# Distribution post-filtering should be nicely centered around 0.5
par(mar=c(4,1,1,2))
for (i in 1:ncol(HQfreqsomNA)){
    hist(HQfreqsomNA[,i], main=length(na.omit(HQfreqsomNA[,i])), breaks=100, xlab=sample_names[i], cex.main=0.3, cex.lab=0.5, cex.axis=0.5)
}
sum1 = apply(HQfreqsomNA, MARGIN=2, FUN=function(X) as.numeric(summary(X)))
rownames(sum1) = c("Min.", "1st Qu.",  "Median",    "Mean","3rd Qu.",    "Max.",    "NA's" )
hist(sum1["Median",], breaks=100, main="", xlab="Median VAF", cex.lab=0.5, cex.axis=0.5)
colnames(HQfreqsomNA)[which(sum1["Median",]<0.4)]



```

## Filter out any that look polyclonal
```{r}
dpkeep = sample_names[ (sample_names != colnames(HQfreqsomNA)[which(sum1["Median",]<0.4)] ) ]
HQfreqsomNA = HQfreqsomNA[,colnames(HQfreqsomNA) %in% dpkeep]
Nsamples = length(dpkeep)
sample_names = dpkeep
Nsamples # Number of remaining samples

```


## Reading in cell type metadata
```{r, message=FALSE, warning=FALSE}
metadata = read.csv(paste(groupname,"_sampleinfo.csv", sep=""), stringsAsFactors = FALSE)
  # B Memory    B Naive CD4 Memory  CD4 Naive  CD8 Naive 
  #        4          1          9         33          4 
m1 = merge(metadata, sample_names, by=1, all.y=TRUE)
## Use files that already have correct Cell.type2 columns
#m1$Cell.type[is.na(m1$Cell.type)] = "HSC"
# m1$Cell.type2 = m1$Cell.type
# m1$Cell.type2[m1$Cell.type2=="CD8 Memory"] = "T Memory"
# m1$Cell.type2[m1$Cell.type2=="CD4 Memory"] = "T Memory"
# m1$Cell.type2[m1$Cell.type2=="CD8 Naive"] = "T Naive"
# m1$Cell.type2[m1$Cell.type2=="CD4 Naive"] = "T Naive"
```

## Separating by sample
```{r, message=FALSE, warning=FALSE}
somsample_list2 = list()
goodsamples = sample_names

for (i in 1:length(goodsamples)){
#for (i in 1:Nsamples){
  #ind = 1
  ind=which(colnames(HQfreqsomNA2)==goodsamples[i]) ## for samples 6-10
  focalfreq = data.frame(HQfreqsomNA2[,1:5], HQ.alt.freq=HQfreqsomNA2[,ind])
  somsample_list2[[i]] = na.omit(focalfreq)
}
names(somsample_list2) = goodsamples
```

## Output files of genotypes for tree building
```{r, message=FALSE, warning=FALSE}
## 1 for present, 0 for absent, 3 for missing data
gens = HQfreqsomNA[ , colnames(HQfreqsomNA) %in% goodsamples]
#gens = HQfreqsomNA
gens[is.na(gens)] = 0 
gens[gens>0] = 1
Nshared = apply(gens, MARGIN=1, FUN=sum)
gens_shared2 = gens[Nshared>1, ] # 7966  329
gens_shared3 = gens[Nshared>2, ] # 2328
gens_shared4 = gens[Nshared>3, ] # 1081
gens_non0 = gens[Nshared>=1, ]
write.table(gens_non0, file=paste("genotypes_non0_",groupname,"_",ncol(gens),"samples.txt", sep=""), quote=F, col.names = T, row.names = T) #420803
write.table(gens_shared2, file=paste("genotypes_shared_",groupname,"_",ncol(gens),"samples.txt", sep=""), quote=F, col.names = T, row.names = T) #9625

## filter and include missing 
HQfreqsom3 = HQfreq_keep[ , which(colnames(HQfreq_keep) %in% goodsamples)] # has some NA values
HQaltsom3 = HQalt_keep[ , which(colnames(HQalt_keep) %in% goodsamples)] # has some NA values
tmp2 = data.frame(HQfreqsom3, HQaltsom3)
Ngoodsamples = length(goodsamples)

HQfreqsom3[ (tmp2[, 1:Ngoodsamples]<=0.2 & tmp2[, (Ngoodsamples+1):(2*Ngoodsamples)]>=1) ] = 3  ## "missing if less than 20% but 1 or more alt reads"
HQfreqsom3[ (tmp2[, 1:Ngoodsamples]>0.2 & tmp2[, (Ngoodsamples+1):(2*Ngoodsamples)]>=2) ] = 1  ## Present if 2 or more alt and 20% or greater
HQfreqsom3[ (tmp2[, 1:Ngoodsamples]>0.2 & tmp2[, (Ngoodsamples+1):(2*Ngoodsamples)]==1) ] = 3  ## missing if 2 or more
HQfreqsom3[ tmp2[, (Ngoodsamples+1):(2*Ngoodsamples)]==0 ] = 0  ## Not present if no alt reads
HQfreqsom3[ is.na(HQfreqsom3)] = 0  ## Deal with NA's
#         0         1         3 
# 125209106    399907   2659347 
sum1 = apply(HQfreqsom3, MARGIN=1, FUN=function(X) sum(X==1))
sum2plus = apply(HQfreqsom3, MARGIN=1, FUN=function(X) sum(X==1)>=2 )
HQfreqsom3b = HQfreqsom3[sum2plus, ]  # 9625  379
write.table(HQfreqsom3b, file=paste("scite_genotypes_alt2freq20missing_", groupname, ".txt", sep=""), quote=F, col.names = F, row.names = F)
write.table(rownames(HQfreqsom3b), file=paste("scite_mutations_alt2freq20missing_", groupname, ".txt", sep=""), quote=F, col.names = F, row.names = F)
write.table(colnames(HQfreqsom3b), file=paste("scite_samples_alt2freq20missing_", groupname, ".txt", sep=""), quote=F, col.names = F, row.names = F)
system(paste("Rscript /Users/hm8/sanger/scripts/scitegenformat2phygenformat.R scite_genotypes_alt2freq20missing_", groupname, ".txt", sep="") )
```

## Number of mutations per colony (pre-dp adjustment)
```{r}
mut_per_colony = data.frame(colony=names(somsample_list2), Nmut = unlist(lapply(somsample_list2, FUN=nrow)))
hist(mut_per_colony$Nmut, breaks=50)
```
## Number of mutations vs coverage
```{r}
m2 = merge(m1, mut_per_colony, by=1)
m2$Cell.type2 = factor(m2$Cell.type2, levels=c("HSC","B Naive","B Memory","T Naive","T Memory" ) ) 
m3 = merge(data.frame(colony=names(meandp), meandp), m2, by=1)
ggplot(m3) +
  geom_point(aes(meandp,Nmut,fill=Cell.type2), pch=21) + 
  scale_fill_brewer(type="qual", palette=3) + 
  ylab("# Mutations") +
  xlab("Coverage per colony")+
  ggtitle(groupname)
#ggsave(file=paste("figures/mutload_celltype_", groupname,".pdf", sep=""), width=6,height=4)
```

## Model to correct for coverage
```{r}
### Asymptotic model
#m4 = m3[m3$Cell.type2 == "HSC" ,  ]
m4 = m3[ (m3$Cell.type2 == "B Naive" | m3$Cell.type2 == "T Naive") & m3$meandp<50 ,  ] ## at very high dp, get errors
m4sorted = sortedXyData(m4$meandp, m4$Nmut)
model.as = NLSstAsymptotic(m4sorted)
model.as

mymodel = function(x){
  b0 = model.as[1]
  b1 = model.as[2]
  lrc = model.as[3]
  b0 + b1*(1-exp(-exp(lrc) * x))  
}

myrange=1:round(max(m4$meandp))
asp = data.frame(x=myrange, y=unlist(lapply(myrange, FUN=mymodel)))
ggplot(m4) +
  geom_point(aes(x=meandp,y=Nmut), data=m4, pch=21) + 
  scale_fill_brewer(type="qual", palette=3) + 
  ylab("# Mutations") +
  xlab("Coverage per colony")+
  geom_line(aes(x=x, y=y), data=asp)+
  ggtitle(groupname)
#ggsave(file=paste("mutload_model_adjust_asy_", groupname,".pdf", sep=""), width=4.5,height=4)
```

## Normalize all to a read depth of 20:
```{r}
new_nmut = function(rd, nmut){
  nmut + ( mymodel(20) - mymodel(rd))
}
m3$Nmut_adj_as = apply(m3, MARGIN=1, FUN=function(X)  new_nmut(rd=as.numeric(X[which(colnames(m3)=="meandp")]), nmut=as.numeric(X[which(colnames(m3)=="Nmut")]) ) )
ggplot(m3) +
  geom_point(aes(meandp,Nmut_adj_as,fill=Cell.type2), pch=21) + 
  scale_fill_brewer(type="qual", palette=3) + 
  ylab("# Mutations") +
  xlab("Coverage per colony")+
  ggtitle(groupname)
#ggsave(file=paste("mutload_celltype_adjust_asy_", groupname,".pdf", sep=""), width=6,height=4)
```

